samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -19,36 +19,201 @@
-    // 3. create new instance of StreamApplication that would generate the optimized graph
-    // 4. create all input/output/intermediate topics
-    // 5. create the configuration for StreamProcessor
-    // 6. start the StreamProcessor w/ optimized instance of StreamApplication
+  @Override
+  public void run(StreamApplication app) {
+    try {
+      // 1. initialize and plan
+      ExecutionPlan plan = getExecutionPlan(app);
+
+      // 2. create the necessary streams
+      createStreams(plan.getIntermediateStreams());
+
+      // 3. start the StreamProcessors
+      if (plan.getJobConfigs().isEmpty()) {
+        throw new SamzaException("No jobs to run.");
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+        log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+        StreamProcessor processor = createStreamProcessor(jobConfig, app);
+        processor.addLifeCycleAware(new LocalProcessorListener(processor));
+        processors.add(processor);
+        processor.start();
+      });
+      appStatus = ApplicationStatus.Running;
+
+      // 4. block until the processors are done or there is a failure
+      awaitComplete();
+
+    } catch (Throwable t) {
+      throw new SamzaException("Failed to run application", t);
+    } finally {
+      if (coordination != null) {
+        coordination.reset();
+      }
+    }
   }
 
   @Override
   public void kill(StreamApplication streamApp) {
+    processors.forEach(StreamProcessor::stop);
   }
 
   @Override
   public ApplicationStatus status(StreamApplication streamApp) {
-    return null;
+    return appStatus;
+  }
+
+  /**
+   * Create the Coordination needed by the application runner.
+   * @return {@link CoordinationUtils}
+   * @throws Exception exception
+   */
+  private CoordinationUtils getCoordinationUtils() throws Exception {
+    ApplicationConfig appConfig = new ApplicationConfig(config);
+    String clazz = appConfig.getCoordinationServiceFactoryClass();
+    if (clazz != null) {
+      CoordinationServiceFactory factory = ClassLoaderHelper.fromClassName(clazz);
+      String groupId = String.format("app-%s-%s", config.get(JobConfig.JOB_NAME()), config.get(JobConfig.JOB_ID(), "1"));
+      return factory.getCoordinationService(groupId, String.valueOf(processorId), config);
+    }
+    else {
+      return null;
+    }
+  }
+
+  /**
+   * Create intermediate streams.
+   * @param intStreams intermediate {@link StreamSpec}s
+   * @throws Exception exception
+   */
+  /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
+    if (!intStreams.isEmpty()) {
+      if (coordination != null) {
+        Latch initLatch = coordination.getLatch(1, LATCH_INIT);
+        coordination.getLeaderElector().tryBecomeLeader(() -> {
+          getStreamManager().createStreams(intStreams);
+          initLatch.countDown();
+        });
+        initLatch.await(LATCH_TIMEOUT, TimeUnit.MINUTES);
```

Comment:
Just clarified with Navina that the latch from the Coordinator is shared among the leader and followers. Hence, discard the previous comment.